---
title: PRJNA498500 Analysis
format: html
author: Gennaro Calendo
---

Import and prepare counts and sample metadata

## Set up

```{r}
suppressPackageStartupMessages(library(coriell))
suppressPackageStartupMessages(library(here))
suppressPackageStartupMessages(library(data.table))
suppressPackageStartupMessages(library(ggplot2))
suppressPackageStartupMessages(library(patchwork))
suppressPackageStartupMessages(library(SummarizedExperiment))
suppressPackageStartupMessages(library(edgeR))

datafiles <- here("results", "PRJNA498500", "data-files", "01")
figures <- here("results", "PRJNA498500", "figures", "01")
rdsfiles <- here("results", "PRJNA498500", "rds-files", "01")

invisible(sapply(c(datafiles, figures, rdsfiles), dir.create, showWarnings=F, recursive=T))
```

## Read in `rmskProfiler` counts

```{r}
se <- readRDS(here("data", "PRJNA498500", "03_rmskProfiler", "se.rds"))
rowData(se)$Ranges <- NULL
```

## Read in metadata and align with SE

```{r}
metadata <- fread(here("doc", "PRJNA498500_SraRunTable.csv"))
metadata <- janitor::clean_names(metadata)
```

Align with SE object and add as object metadata

```{r}
setDF(metadata, rownames=metadata$geo_accession_exp)
metadata <- metadata[colnames(se), ]
colData(se) <- cbind(colData(se), metadata)
```

## Remove any TEs within exons 

```{r}
# Select hash IDs in exons to drop
drop_te <- subset(rowData(se), 
  !is.na(Hash) & (hasUnstrandedExonic == TRUE | hasUnstranded3UTR == TRUE | hasUnstranded5UTR == TRUE),
  select = "Hash",
  drop = TRUE
  )

se <- se[!rownames(se) %chin% drop_te, ]
```

## Drop outlier samples

By library density and boxplots

```{r}
se <- se[, !colnames(se) %in% c("GSM3447011", "GSM3447019")]
```

## `edgeR` pipeline

This study does not actually have response RECIST categories. They simply report:

>multi institution clinical trial to evaluate immune responses and survival following neoadjuvant and/or adjuvant therapy with pembrolizumab, a programmed cell death protein 1 (PD-1) monoclonal antibody, in 35 patients with recurrent, surgically resectable glioblastoma. Patients who were randomized to receive neoadjuvant pembrolizumab, with continued adjuvant therapy following surgery, had significantly extended overall survival compared to patients that were randomized to receive adjuvant, post-surgical PD-1 blockade alone (hazard ratio = 0.39; P = 0.04, log-rank test). Neoadjuvant PD-1 blockade was associated with upregulation of T cell and interferon-Î³-related genes, but downregulation of cell cycle related genes within the tumor, which was not seen in patients that received adjuvant therapy alone.

### Filter for expression

```{r}
y <- SE2DGEList(se)
y$samples$group <- y$samples$therapy

keep <- filterByExpr(y)
y <- y[keep,,keep.lib.sizes=FALSE]
```

### Compute TMM scaling factors

```{r}
y <- normLibSizes(y)
```

Show the MDS plot of the normalized counts. 

```{r}
png(here(figures, "mds.png"), width=6, height=5, units="in", res=300)
plotMDS(y, col=factor(y$samples$group), pch=19, main="By Response")
legend("bottomleft", legend=c("adjuvant pembrolizumab", "neoadjuvant pembrolizumab"), col=palette()[1:2], pch=19)
dev.off()
```

Show QC plots

```{r}
lcpm <- cpm(y, log=TRUE)

coriell::plot_boxplot(lcpm, metadata=y$samples, fillBy="therapy", rle=TRUE)
ggsave(here(figures, "rle-boxplot.png"), width=9, height=6)

coriell::plot_density(lcpm, metadata=y$samples, colBy="therapy")
ggsave(here(figures, "lib-density.png"), width=9, height=6)
```

### Perform DE testing

```{r}
design <- model.matrix(~group, data=y$samples)
fit <- glmQLFit(y, design=design, robust=TRUE)

# Test for differential expression
na_vs_a <- glmQLFTest(fit, coef=2) 

# Combine all results into a list of data.frames
results <- list("neoadjuvant_vs_adjuvant" = edger_to_df(na_vs_a))
```

Show differential expression plot

```{r}
# Transcripts
v1 <- plot_volcano(subset(results[[1]], feature_id %like% "^ENST")) + 
  labs(title = "Neoadjuvant vs Adjuvant")
m1 <- plot_md(subset(results[[1]], feature_id %like% "^ENST")) + 
  labs(title =  "Neoadjuvant vs Adjuvant")

(v1 | m1)
ggsave(here(figures, "differential-expression_tx.png"), width=12, height=6)

# Transcripts
v1 <- plot_volcano(subset(results[[1]], !feature_id %like% "^ENST")) + 
  labs(title = "Neoadjuvant vs Adjuvant")
m1 <- plot_md(subset(results[[1]], !feature_id %like% "^ENST")) + 
  labs(title =  "Neoadjuvant vs Adjuvant")

(v1 | m1)
ggsave(here(figures, "differential-expression_te.png"), width=12, height=6)
```

## Save the final SE object for meta-analysis

```{r}
processed <- SummarizedExperiment(
    assays = list(counts = y$counts, logcounts = lcpm),
    colData = y$samples,
    rowData = y$genes,
    metadata = list(results=results, fit=fit, design=design)
)

saveRDS(processed, here(rdsfiles, "processed-se.rds"))
```

