---
title: PRJNA762343 Analysis
format: html
author: Gennaro Calendo
---

Import and prepare counts and sample metadata

## Set up

```{r}
suppressPackageStartupMessages(library(coriell))
suppressPackageStartupMessages(library(here))
suppressPackageStartupMessages(library(data.table))
suppressPackageStartupMessages(library(ggplot2))
suppressPackageStartupMessages(library(SummarizedExperiment))
suppressPackageStartupMessages(library(edgeR))

datafiles <- here("results", "PRJNA762343", "data-files", "01")
figures <- here("results", "PRJNA762343", "figures", "01")
rdsfiles <- here("results", "PRJNA762343", "rds-files", "01")

invisible(sapply(c(datafiles, figures, rdsfiles), dir.create, showWarnings=F, recursive=T))
```

## Read in `rmskProfiler` counts

```{r}
se <- readRDS(here("data", "PRJNA762343", "03_rmskProfiler", "se.rds"))
rowData(se)$Ranges <- NULL
```

## Read in metadata and align with SE

```{r}
metadata <- fread(here("doc", "PRJNA762343_SraRunTable.csv"))
metadata <- janitor::clean_names(metadata)
```

They did not include clinical data in run info. Read in from GEO and supp info

```{r}
clinical <- readxl::read_excel(here("doc", "GSE183924_Clinical_Data_with_CIBERSORT_analysis_for_GEO.xlsx"))
clinical <- janitor::clean_names(clinical)

metadata <- merge(
    x = metadata,
    y = clinical,
    by.x = "patient_id",
    by.y = "patient_id",
    all.x = TRUE
    )

setDF(metadata, rownames=metadata$geo_accession_exp)
metadata <- metadata[colnames(se), ]
colData(se) <- cbind(colData(se), metadata)
```

## Remove any TEs within exons 

```{r}
# Select hash IDs in exons to drop
drop_te <- subset(rowData(se), 
  !is.na(Hash) & (hasUnstrandedExonic == TRUE | hasUnstranded3UTR == TRUE | hasUnstranded5UTR == TRUE),
  select = "Hash",
  drop = TRUE
  )

se <- se[!rownames(se) %chin% drop_te, ]
```

## Remove outlier samples

By RLE boxplot and densities

```{r}
outliers <- paste0("GSM55744", c(38, 46, 48, 55, 57))

se <- se[, !colnames(se) %in% outliers]
```

## `edgeR` pipeline

**This experiment may be better off modelled as expr ~ months** rather than rfs censor vs not

### Filter for expression

```{r}
y <- SE2DGEList(se)

# Relapse or censor == 1, make no relapse the test condition
y$samples$group <- factor(y$samples$rfs_censor, levels=c("1", "0"))

keep <- filterByExpr(y)
y <- y[keep,,keep.lib.sizes=FALSE]
```

### Compute TMM scaling factors

```{r}
y <- normLibSizes(y)
```

Show the MDS plot of the normalized counts. 

```{r}
png(here(figures, "mds.png"), width=6, height=5, units="in", res=300)
plotMDS(y, col=y$samples$group, pch=19, main="By RFS Censor")
legend("topright", legend=c("relapse", "no relapse"), col=palette()[1:2], pch=19)
dev.off()
```

Show QC plots

```{r}
lcpm <- cpm(y, log=TRUE)

coriell::plot_boxplot(lcpm, metadata=y$samples, fillBy="group", rle=TRUE)
ggsave(here(figures, "rle-boxplot.png"), width=9, height=6)

coriell::plot_density(lcpm, metadata=y$samples, colBy="group")
ggsave(here(figures, "lib-density.png"), width=9, height=6)
```

### Perform DE testing

```{r}
design <- model.matrix(~group, data=y$samples)
fit <- glmQLFit(y, design=design, robust=TRUE)

# Test for differential expression
r_vs_nr <- glmQLFTest(fit, coef=2) 

# Combine all results into a list of data.frames
results <- list("relapse_vs_norelapse" = edger_to_df(r_vs_nr))
```

## Save the final SE object for meta-analysis

```{r}
processed <- SummarizedExperiment(
    assays = list(counts = y$counts, logcounts = lcpm),
    colData = y$samples,
    rowData = y$genes,
    metadata = list(results=results, fit=fit, design=design)
)

saveRDS(processed, here(rdsfiles, "processed-se.rds"))
```
